### PR TITLE
Early exit opportunity for stakeholders

### PR DESCRIPTION
**Pull request by wilhempujar**
This PR is inspired by https://github.com/Stacktical/stacktical-dsla-contracts/pull/134 and introduces:
* A new formula for provider rewards
* A check to ensure that there is always liquidity available for payouts, derived from the SLI deviation cap
* A reward of 25% for equal / non equal SLO types
* A mechanism locking stakeholders for at least 1 verification and an early exit opportunity
* Better revert messages
* Better variable names
* Updated tests

### CODE REVIEW

File: contracts/Staking.sol [hunk 1]
```diff
@@ -290,12 +295,24 @@ contract Staking is Ownable, ReentrancyGuard {
...
-            uint256 reward = usersStake.mul(_rewardPercentage).div(_precision);
+
+            uint256 reward = providersPool[tokenAddress]
+                .div(leverage)
+                .mul(_rewardPercentage)
+                .div(_precision);
+
+            // Reward must be less than 25% of usersPool to ensure payout at all time
+            if (reward > usersPool[tokenAddress].mul(25).div(100)) {
+                reward = usersPool[tokenAddress].mul(_rewardPercentage).div(
```
Comment (Axel-CH) [inline contracts/Staking.sol:306]: I like the idea of having a capped provider rewards but I think this line could cause reward instability. Are you sure that if the reward cap is activated I won't cause a reward drop bigger than if the reward is almost capped?
Why not setting a capped reward at `usersPool[tokenAddress].mul(25).div(100)` ?

File: contracts/Staking.sol [hunk 2]
```diff
@@ -316,14 +333,20 @@ contract Staking is Ownable, ReentrancyGuard {
...
-            uint256 compensation = usersStake
+            uint256 compensation = usersPool[tokenAddress]
                 .mul(leverage)
                 .mul(_rewardPercentage)
                 .div(_precision);
 
-            providerPool[tokenAddress] = providerPool[tokenAddress].sub(
+            // Compensation must be less than 25% of providersPool to ensure payout at all time
+            if (compensation > providersPool[tokenAddress].mul(25).div(100)) {
+                compensation = providersPool[tokenAddress]
```
Comment (Axel-CH) [inline contracts/Staking.sol:344]: Same interrogation, why not setting compensation at `providersPool[tokenAddress].mul(25).div(100)` to prevent instability?

File: contracts/Staking.sol [hunk 3]
```diff
@@ -344,27 +367,33 @@ contract Staking is Ownable, ReentrancyGuard {
...
+                lastProviderStake[msg.sender] < _nextVerifiablePeriod,
+                'Provider lock-up until the next verification.'
+            );
 
-        require(
-            providerPool[_tokenAddress].sub(_amount) >=
-                usersPool[_tokenAddress].mul(leverage),
-            'Withdrawal exceeds leveraged cap.'
-        );
+            // Allow provider withdrawal as long as the provider user pool exceeds the provider pool
```
Comment (Axel-CH) [inline contracts/Staking.sol]: Small wording mistake should be `Allow provider withdrawal as long as the provider pool exceeds the user pool`

File: contracts/Staking.sol [hunk 4]
```diff
@@ -19,8 +19,8 @@ contract Staking is Ownable, ReentrancyGuard {
     using SafeERC20 for IERC20;
 
     enum Position {
```
Comment (Axel-CH) [inline contracts/Staking.sol:21]: I get the Idea to move away from trading specific semantic but as you know i'm not a fan of the OK/KO pair as it is removing the notion of temporality (expected future result)

Comment (wilhempujar) [reply contracts/Staking.sol:21]: I agree, but that's exactly what we need to remove here :)

File: contracts/Staking.sol (hunk 1 above)
Comment (wilhempujar) [reply contracts/Staking.sol:306]: Because _rewardPercentage is already capped at 25% is the SLORegistry

So at the end of the day, the reward is `25% / leverage * providersPool` if there is enough liquidity in usersPool, or simply **up to** 25% of the usersPool.

**Review by Axel-CH — COMMENTED** [review]
We must validate that the new pool size related reward cap don't introduce reward or compensation instability when activated.

File: contracts/Staking.sol (hunk 2 above)
Comment (wilhempujar) [reply contracts/Staking.sol:344]: See my other answer, that's also because _rewardPercentage is already capped at 25% is the SLORegistry

Here the compensation is `25% * leverage * usersPool` if there is enough liquidity in providersPool, or simply **up to** 25% of the providersPool

File: contracts/Staking.sol (hunk 3 above)
Comment (wilhempujar) [reply contracts/Staking.sol]: Fixed!

File: contracts/Staking.sol (hunk 1 above)
Comment (wilhempujar) [reply contracts/Staking.sol:306]: Based on our discussions there are two main approaches:
* A/ Apply the 25% cap to the usersPool
* B/ Apply the SLI deviation, capped to 25%, to the usersPool

I think A is more of a purist approach, than something that can be market-validated. 

From a user perspective, it's better to lose less, and have a longstanding coverage (through the availability of deposited premiums). I am strongly assuming they will prefer approach B, and our data also shows this.

I'm vouching for keeping the user-centric approach implemented here. At least until v2.0

Comment (Axel-CH) [reply contracts/Staking.sol:306]: Good, let's resolve this. Specific IT test cases should be added in DTK to address these edge cases.

**Review by Axel-CH — APPROVED** [review]
LGTM, Edge cases involving Pools should be tested in DTK

**Review by gmspacex — APPROVED** [review]
LGTM, Just pushed some optimizations.
SLARegistry contract size reduced to 24.395 KB now.
